File: jsrc/exceptions/SxxMachine/JavaException.java
package com.googlecode.prolog_cafe.lang;
/**
 * Java error.<br>
 * There will be a Java error when 
 * a Java exception is threw during interoperating with Java in Prolog Cafe.
 * The class <code>JavaException</code> wraps a subclass of <code>java.lang.Exception</code>.
 *
 * @author Mutsunori Banbara (banbara@kobe-u.ac.jp)
 * @author Naoyuki Tamura (tamura@kobe-u.ac.jp)
 * @version 1.0
 */
public class JavaException extends BuiltinException {

	private static final long serialVersionUID = -7510890457700984457L;

	/** A functor symbol of <code>java_error/3</code>. */
    public static final SymbolTerm JAVA_ERROR = SymbolTerm.intern("java_error", 3);

    /** Holds a Java exception. */
    public Exception e;

    /** Constructs a new <code>JavaException</code> with a Java exception. */
    public JavaException(Exception _e) {
    	initCause(_e);
    	e = _e;
    }

    /** Constructs a new <code>JavaException</code> with the given arguments. */
    public JavaException(Operation _goal, int _argNo, Exception _e) {
    	this(_e);
    	this.goal    = _goal;
    	this.argNo   = _argNo;
    }

    /** Returns a term representation of this <code>JavaException</code>:
     * <code>java_error(goal,argNo,exception)</code>.
     */
    public Term getMessageTerm() {
		Term[] args = {
			(goal==null)?SymbolTerm.create("<Goal unknown>"):new JavaObjectTerm(goal), 
		    new IntegerTerm(argNo),
		    new JavaObjectTerm(e)};
		return new StructureTerm(JAVA_ERROR, args);
    }

    /** Returns a underlying Java exception. */
    public Exception getException() {
    	return e;
    }

    /** Returns a string representation of this <code>JavaException</code>. */
    public String toString() {
		StringBuilder sb = new StringBuilder("{JAVA ERROR: ");
		sb.append(goal);
		if (argNo>0){
			sb.append(" - arg ").append(argNo);
		}
		sb.append(", occurs ").append(e);
		sb.append("}");
		return sb.toString();
    }
}

File: src/exceptions/SxxMachine/JavaException.cpp
using namespace std;

#include "JavaException.h"
#include "../../machine/SxxMachine/SymbolTerm.h"
#include "../../machine/SxxMachine/Operation.h"
#include "../../machine/SxxMachine/Term.h"
#include "../../machine/SxxMachine/TermData.h"

namespace SxxMachine
{

SymbolTerm *const JavaException::JAVA_ERROR = SymbolTerm::intern("java_error", 3);

	JavaException::JavaException(runtime_error _e) : e(_e)
	{
		initCause(_e);
	}

	JavaException::JavaException(Operation _goal, int _argNo, runtime_error _e) : JavaException(_e)
	{
		this->goal = _goal;
		this->argNo = _argNo;
	}

	Term *JavaException::getMessageTerm()
	{
		std::vector<Term*> args = {(this->goal == nullptr)?SymbolTerm::create("<Goal unknown>"):TermData::FFIObject(this->goal), TermData::Integer(this->argNo), TermData::FFIObject(this->e)};
		return TermData::createErrorTerm(this, JAVA_ERROR, args);
	}

	runtime_error JavaException::getException()
	{
		return this->e;
	}

	wstring JavaException::toString()
	{
		StringBuilder *sb = new StringBuilder("{JAVA ERROR: ");
		sb->append(this->goal);
		if (this->argNo > 0)
		{
			sb->append(" - arg ")->append(this->argNo);
		}
		sb->append(", occurs ")->append(this->e);
		sb->append("}");
		return sb->toString();
	}

	wstring JavaException::getMessage()
	{
	return toString();
	}
}
